### PR TITLE
Use Netty ByteBuf Bulk Operations for Faster Deserialization

### PR DESCRIPTION
**Pull request by original-brownbear**
A few optimizations to the deserialization from Netty `ByteBuf` that leverage Netty's bulk read APIs:
* Use bulk methods to read numeric types to save bounds checks etc.
* The duplication between NIO and Netty codebases obviously gets even worse with this, but I went with the existing approach to things for now
* Similar to #39634

### CODE REVIEW

**Comment by jpountz** [comment]
Have you been able to measure any speedup?

**Comment by original-brownbear** [comment]
@jpountz I ran some quick and dirty benchmarks (I can do better ones if desired):

```
public class DirtyBenchmark {

    public static void main(String[] args) throws IOException {
        final ByteBuf byteBuf = PooledByteBufAllocator.DEFAULT.directBuffer(512);
        final String first = "some string value";
        try (OutputStreamStreamOutput outputStream = new OutputStreamStreamOutput(new ByteBufOutputStream(byteBuf))) {
            for (int i = 0; i < 10; ++i) {
                outputStream.writeInt(2);
                outputStream.writeLong(Long.MAX_VALUE);
                outputStream.writeString(first);
            }
        }
        final long start = System.currentTimeMillis();
        final BytesReference bytesReference = Netty4Utils.toBytesReference(byteBuf);
        for (int i = 0; i < 20_000_000; ++i) {
            byteBuf.readerIndex(0);
            for (int j = 0; j < 10; ++j) {
                final StreamInput input = bytesReference.streamInput();
                if (input.readInt() != 2) {
                    System.exit(1);
                }
                if (input.readLong() != Long.MAX_VALUE) {
                    System.exit(1);
                }
                if (first.equals(input.readString()) == false) {
                    System.exit(1);
                }
            }
        }
        System.out.println(System.currentTimeMillis() - start);
    }
}
```

for this one it's about twice as fast in terms of throughput with fewer allocations and less GC time.
Obviously, the shorter the messages the smaller the improvement (if I just do `int`, `long`, `String` instead of the 10x repeat the runtime is exactly the same pretty much but it still needs less GC thanks to saving the `char[]`).

**Comment by s1monw** [comment]
I don't like optimizing this on the netty end. If we want to make this more efficient lets make the write side different such that we can read bytes directly instead. I think there are several options like for instance we can change the way we write the chunks and mark the last chunk of the string with a trailing bit of it's chunksize being 0 (similar to how vint works) this way we can read in chunk and read from the stream in bytes and then use Unicodeutils to convert from `byte[]` to `char[]` the sting format would look like this`<size_in_char,chunk_size,byte_chunk,[chunk_size,byte_chunk]>` It might be a bit tricky to make sure we never write a chunk that ends at a supplementary character but that's doable. This way all implementations will have the optimization and we get much better test coverage.

WDYT?

**Comment by original-brownbear** [comment]
@s1monw I like that idea, but maybe that's something for another PR? (I can't look into that right away).
I removed the string changes now but kept the numeric read changes. This still gives a significant speedup of about 50% in the dirty benchmark from [above](https://github.com/elastic/elasticsearch/pull/40158#issuecomment-474313736) which seems to come from better inlining since it wasn't visible when running with a profiler. I think for primitive reads it's not too dirty to use the bulk methods, those simply are very implementation dependent?

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java [hunk 1]
```diff
@@ -1064,7 +1063,7 @@ public static StreamInput wrap(byte[] bytes, int offset, int length) {
      * Reads a vint via {@link #readVInt()} and applies basic checks to ensure the read array size is sane.
      * This method uses {@link #ensureCanReadBytes(int)} to ensure this stream has enough bytes to read for the read array size.
      */
-    private int readArraySize() throws IOException {
+    protected int readArraySize() throws IOException {
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java]: is this necessary?

**Review by s1monw — APPROVED** [review]
+1 to another PR

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java]: Nah it was just for the string optimization -> reverted :)

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/bwc

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/bwc

**Comment by original-brownbear** [comment]
@s1monw thanks!

**Pull request merged**